### ISSUE TITLE
Show better warning for trying to cast non-u8 scalar to char

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust#44201]
```rust
fn main() {
    const XYZ: char = 0xEF8888 as char;
    println!("{}", XYZ);
}
```

[Will warn "literal is out of range for `u8`"](https://play.rust-lang.org/?gist=d52387ce80384df52aedf3a2669dfa74&version=nightly) because `u8` is the only `as char` cast that is acceptable.

We should show a better warning and suggest the `\u` syntax instead.

**Comment by estebank** [comment]
In the following code:
https://github.com/rust-lang/rust/blob/442b7bd10ab1955832fac39e47671327747033c6/src/librustc_lint/types.rs#L165-L182

check if `e.hir_id` parent expr is a cast to `char` and emit a `help` explaining that only `u8` can be casted to `char` and a suggestion to write `const XYZ: char = '\u{EF8888}';` instead.
